kmp-sharing/src/androidMain/kotlin/com/swmansion/kmpsharing/Sharing.kt

```diff
@@ -0,0 +1,99 @@
+
+import android.annotation.SuppressLint
+import android.content.Context
+import android.content.Intent
+import androidx.core.content.FileProvider
+import androidx.core.net.toUri
+import java.io.File
+import java.net.URLConnection
+
+public actual class Sharing {
+    public actual companion object {
+        public actual fun share(url: String, options: SharingOptions?) {
+            val context = getContext()
+
+            try {
+                val file = getLocalFileFromUrl(url)
+
+                val contentUri =
+                    FileProvider.getUriForFile(context, "${context.packageName}.fileprovider", file)
+
+                val mimeType =
+                    options?.androidMimeType
+                        ?: URLConnection.guessContentTypeFromName(file.name)
+                        ?: "*/*"
+
+                val intent =
+                    Intent(Intent.ACTION_SEND).apply {
+                        putExtra(Intent.EXTRA_STREAM, contentUri)
+                        setTypeAndNormalize(mimeType)
+                        addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
+
+                        options?.androidDialogTitle?.let { title ->
+                            putExtra(Intent.EXTRA_TEXT, title)
+                        }
+                    }
+
+                val chooser = Intent.createChooser(intent, options?.androidDialogTitle ?: "Share")
+                chooser.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
+
+                context.startActivity(chooser)
+            } catch (e: Exception) {
+                throw SharingFailedException("Failed to share: ${e.message}", e)
+            }
+        }
+
+        @SuppressLint("PrivateApi")
+        private fun getContext(): Context {
+            return try {
+                val applicationClass = Class.forName("android.app.ActivityThread")
+                val currentApplication =
+                    applicationClass.getMethod("currentApplication").invoke(null) as? Context
+
+                if (currentApplication != null) {
+                    currentApplication
+                } else {
+                    val activityThread =
+                        applicationClass.getMethod("currentActivityThread").invoke(null)
+                    val app =
+                        activityThread
+                            ?.javaClass
+                            ?.getMethod("getApplication")
+                            ?.invoke(activityThread) as? Context
+
+                    app ?: throw IllegalStateException("Could not get application context")
+                }
+            } catch (e: Exception) {
+                throw IllegalStateException("Could not get Android context automatically.", e)
+            }
+        }
+
+        @Throws(SharingInvalidArgumentException::class)
+        private fun getLocalFileFromUrl(url: String?): File {
+            if (url == null) {
+                throw SharingInvalidArgumentException("URL to share cannot be null.")
+            }
```

Comment:
https://kotlinlang.org/api/core/kotlin-stdlib/kotlin/require.html and https://kotlinlang.org/api/core/kotlin-stdlib/kotlin/require-not-null.html. `IllegalArgumentException` is an idiomatic exception class for such cases.